### PR TITLE
Fix `test_gpu_metrics` failure

### PR DESCRIPTION
**Pull request by jakirkham**
Fix https://github.com/dask/distributed/issues/4153

The `test_gpu_metrics` test fails. This attempts to fix it.

cc @quasiben

### CODE REVIEW

File: distributed/tests/test_gpu_metrics.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 @gen_cluster()
 async def test_gpu_metrics(s, a, b):
-    from distributed.diagnostics.nvml import handles
+    from distributed.diagnostics.nvml import _pynvml_handles as handles
 
     assert "gpu" in a.metrics
     assert len(s.workers[a.address].metrics["gpu"]["memory-used"]) == len(handles)
```
Comment (jakirkham) [inline distributed/tests/test_gpu_metrics.py:12]: When running this, I encounter another issue. Short snippet from the failure included below for brevity. Not sure if this is an indication of a bug or another issue in the test.

```python
    @gen_cluster()
    async def test_gpu_metrics(s, a, b):
        from distributed.diagnostics.nvml import _pynvml_handles as handles
    
        assert "gpu" in a.metrics
>       assert len(s.workers[a.address].metrics["gpu"]["memory-used"]) == len(handles)
E       TypeError: object of type 'int' has no len()
```

**Comment by quasiben** [comment]
Perhaps we should remove this test.  Most of this is covered in https://github.com/dask/distributed/blob/master/distributed/diagnostics/tests/test_nvml.py .  Thoughts ?

File: distributed/tests/test_gpu_metrics.py (hunk 1 above)
Comment (jakirkham) [reply distributed/tests/test_gpu_metrics.py:12]: We only return one handle from `_pynvml_handles` as well. So I'm not sure if `len` was suppose to be here or if this is a typo.

https://github.com/dask/distributed/blob/a1dc5f437b39c1b35a9b05cbc048e3a793b89715/distributed/diagnostics/nvml.py#L29-L30

Comment (jakirkham) [reply distributed/tests/test_gpu_metrics.py:12]: It looks like this function use to return a `list`, but PR ( https://github.com/dask/distributed/pull/3810 ) changed it to return a single value. So maybe this test doesn't make sense (at least how it is written) any more.

Comment (quasiben) [reply distributed/tests/test_gpu_metrics.py:12]: I changed the way nvml/handles work -- we are now only handing back one handle per worker

**Comment by quasiben** [comment]
We could also add a check for `memory-used` in `test_nvml.py`

**Comment by jakirkham** [comment]
Yeah I'm fine dropping it. After doing a bit of digging (really not familiar with this piece of code), looks like it is just a historical artifact.

Also happy to consolidate into our more recent working tests.

**Comment by jakirkham** [comment]
Ok this now passes for me locally.

**Comment by quasiben** [comment]
Cool!  Maybe let's leave this as the solution and check back in with @jacobtomlinson when he returns before removing

**Comment by jakirkham** [comment]
Looks like the tests in [`distributed/diagnostics/tests/test_nvml.py`]( https://github.com/dask/distributed/blob/6efd5b6147d29ec613bd83c72cdbd6a5a45a1438/distributed/diagnostics/tests/test_nvml.py ) also don't pass for the same reason.

```python
____________________________ test_1_visible_devices ____________________________

    def test_1_visible_devices():
        os.environ["CUDA_VISIBLE_DEVICES"] = "0"
        output = nvml.one_time()
>       assert len(output["memory-total"]) == 1
E       TypeError: object of type 'int' has no len()

distributed/diagnostics/tests/test_nvml.py:21: TypeError
```

**Comment by jakirkham** [comment]
Ok have now fixed that error too.

**Comment by jakirkham** [comment]
> Cool! Maybe let's leave this as the solution and check back in with @jacobtomlinson when he returns before removing

Well the test is still there. Have just moved it to join the other NVML tests. Thought this is what was being suggested.

Mainly my interest here is to get Dask-CUDA's CI to stop failing. Don't have strong feelings about what we do in the future, but maybe we can get this in to fix those issues?

File: distributed/diagnostics/tests/test_nvml.py [hunk 2]
```diff
@@ -17,11 +18,15 @@ def test_one_time():
...
-    assert len(output["memory-total"]) == 1
+    h = nvml._pynvml_handles()
+    assert output["memory-total"] == pynvml.nvmlDeviceGetMemoryInfo(h).total
 
 
 @pytest.mark.parametrize("CVD", ["1,0", "0,1"])
 def test_2_visible_devices(CVD):
+    if pynvml.nvmlDeviceGetCount() <= 1:
+        pytest.skip("Machine only has a single GPU")
+
```
Comment (jakirkham) [inline distributed/diagnostics/tests/test_nvml.py:29]: Added this to skip the multi-GPU test on single GPU machines based on our conversation offline, can you please take a look and see if I got this right, Ben? :)

**Comment by quasiben** [comment]
Thanks @jakirkham.  All this passes now including the other GPU tests within distributed on a DGX machine.  I am getting a failure on a workstation (I think it's a GPU configuration issue:

```python
ipdb> pynvml.nvmlDeviceGetSerial(h)
*** pynvml.nvml.NVMLError_NotSupported: Not Supported
```

I'll look at this early tomorrow morning.

**Comment by jakirkham** [comment]
Yeah looking at [the docs]( https://docs.nvidia.com/deploy/nvml-api/group__nvmlDeviceQueries.html#group__nvmlDeviceQueries_1g56b0c288a1c2eb60b9949aff60d64486 ) I see the following...

> For all products with an inforom.

An [earlier version of the docs]( http://developer.download.nvidia.com/compute/DevZone/NVML/doxygen/group__group4.html#g56b0c288a1c2eb60b9949aff60d64486 ) says...

> For Fermi-class products only.

So maybe this GPU is not supported for this operation? If so, maybe there is an alternative to this API call that would work more generally?

Though it's also possible there's some configuration issue we don't understand (as you have noted). Might be worth verify whether this specific GPU can run this function or not before digging in too deep.

**Comment by quasiben** [comment]
After reviewing/changing [permissions from pynvml](https://github.com/gpuopenanalytics/pynvml#nvml-permissions)  everything works as expected.  Merging in.

Thanks @jakirkham !

**Comment by jakirkham** [comment]
Ah ok. Good to know. Thanks for checking Ben! 😄

**Comment by jacobtomlinson** [comment]
Thanks for dealing with this folks!

**Pull request merged**